mlos_bench/mlos_bench/optimizers/mlos_core_optimizer.py

```diff
@@ -99,26 +95,42 @@ def bulk_register(self,
                       configs: Sequence[dict],
                       scores: Sequence[Optional[Dict[str, TunableValue]]],
                       status: Optional[Sequence[Status]] = None) -> bool:
+
         if not super().bulk_register(configs, scores, status):
             return False
+
         df_configs = self._to_df(configs)  # Impute missing values, if necessary
-        df_scores = pd.Series(
-            [self._extract_target(score) for score in scores],
-            dtype=float) * self._opt_sign
+
+        df_scores = self._adjust_signs_df(
+            pd.DataFrame([{} if score is None else score for score in scores]))
+
+        opt_targets = list(self._opt_targets)
         if status is not None:
+            # Select only the completed trials, set scores for failed trials to +inf.
             df_status = pd.Series(status)
-            df_scores[df_status != Status.SUCCEEDED] = float("inf")
+            df_scores.loc[df_status != Status.SUCCEEDED, opt_targets] = float("inf")
             df_status_completed = df_status.apply(Status.is_completed)
             df_configs = df_configs[df_status_completed]
             df_scores = df_scores[df_status_completed]
-        self._opt.register(df_configs, df_scores)
+
+        # Pass opt_targets columns as scores, and the rest as context.
+        opt_context = set(df_scores.columns).difference(opt_targets)
```

Comment:
Disagree with this one for now.

There's so many things we might collect as metrics in the scores that aren't context that don't think we should default to that in this way.

For instance, imagine I collect throughput, latency, cost, elapsed time, and system metrics (CPU, disk, etc.).
In one run, I may want to optimize for increased throughput and lower elapsed time, where as in another run I may want to reduce CPU metrics and cost.
That data could be reused between experiments to prewarm an optimizer, but the omitted columns from one side vs. another don't tell me anything about the "context" of the test, only its results.

In my mind context are things about the execution environment (e.g., VM core count, VM RAM size, etc.) and the workload identification (WIP), but it wouldn't include execution metrics.

Let's leave this part out and make context handling something explicit in the future.